python/ray/data/tests/test_dataset.py

```diff
@@ -5517,6 +5517,16 @@ def test_ragged_tensors(ray_start_regular_shared):
     ]
 
 
+def test_legacy_do_write(ray_start_regular_shared, tmp_path):
+    out = CSVDatasource()
+    ds = ray.data.range(100).fully_executed()
+    blocks, metadata = zip(*ds._plan.execute().get_blocks_with_metadata())
+    out.do_write(blocks, metadata, tmp_path, ds._uuid)
```

Comment:
This isn't testing that we support do_write() through the Dataset API. For proper backwards compatibility, we need to (1) check if the datasource doesn't have direct write implemented, and (2) in that case, Dataset needs to fallback to using do_write() as an AllToAll stage instead of a OneToOne.

Reply:
I'm checking the presence of "write" attribute to decide whether this should fallback to the legacy path.